### PR TITLE
Abstract storage functionality

### PR DESCRIPTION
**Pull request by dolhuis**
Storage handling has been collected in the GOB_Storage_Handler, and
removed from all the places where storage functionality has been
handled.

Compare and Update have been brought under test

Storage handling can be extracted into GOB-Core (and should be tested
there).

### CODE REVIEW

File: src/gobuploadservice/compare.py [hunk 1]
```diff
@@ -35,39 +24,27 @@ def compare(msg):
...
-        entity = session.query(DbEntity).filter_by(_source=metadata.source,
-                                                   _source_id=entity_id,
-                                                   _date_deleted=None).one_or_none()
-
-        modifications = get_modifications(entity, data, metadata.model)
-        event = get_event_for(entity, data, metadata, modifications)
-        events.append(event)
-
-    session.close()
+    db = GOBStorageHandler(metadata)
```
Comment (Vegter) [inline src/gobuploadservice/compare.py]: Just a small question: is storage not a better (more abstract) name than db?

File: src/gobuploadservice/storage/storage_handler.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    return wrapper_decorator
+
+
+class GOBStorageHandler():
+    """Metadata aware Storage handler """
+    EVENT_TABLE = "event"
+
+    def __init__(self, gob_metadata):
+        """Initialize DBHandler with gob metadata
+
```
Comment (Vegter) [inline src/gobuploadservice/storage/storage_handler.py]: StorageHandler

File: src/Dockerfile [hunk 3]
```diff
@@ -1,6 +1,10 @@
 FROM amsterdam/docker_python:latest
 MAINTAINER datapunt@amsterdam.nl
 
+RUN pip3 install --upgrade pip
+RUN mkdir ~/.ssh
```
Comment (Vegter) [inline src/Dockerfile]: I would recommend including some comments why this is required

File: src/gobuploadservice/__main__.py [hunk 4]
```diff
@@ -6,19 +6,19 @@
...
         'queue': "gob.workflow.request",
-        'handler': compare,
+        'handler': compare.compare,
         'report_back': 'fullupdate.proposal',
         'report_queue': 'gob.workflow.proposal'
     },
     'fullupdate.request': {
         'queue': "gob.workflow.request",
-        'handler': full_update,
+        'handler': update.full_update,
```
Comment (Vegter) [inline src/gobuploadservice/__main__.py]: Nicely anticipating single uploads!

File: src/gobuploadservice/compare.py [hunk 5]
```diff
@@ -35,39 +24,27 @@ def compare(msg):
...
+        current_ids = db.get_current_ids()
+
+        # find deletes by comparing current ids to new entities
+        # if a current_id is not found in the new_entities it is interpreted as a deletion
+        deleted = {current._source_id: None for current in current_ids if current._source_id not in new_entities}
+
+        # combine new and deleted into one set
+        all = {**new_entities, **deleted}
+
+        events = []
```
Comment (Vegter) [inline src/gobuploadservice/compare.py]: Probably add some comments about what is done in the next lines

File: src/gobuploadservice/storage/storage_handler.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
+import functools
```
Comment (Vegter) [inline src/gobuploadservice/storage/storage_handler.py]: Docstrings missing for module introduction and explanation

File: src/gobuploadservice/storage/storage_handler.py [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+from sqlalchemy.orm import Session
+
+from gobuploadservice.config import GOB_DB
+from gobuploadservice.storage.db_models import get_column
+from gobuploadservice.storage.db_models.event import EVENTS, build_db_event
+from gobuploadservice.storage.db_models.metadata import FIXED_COLUMNS, METADATA_COLUMNS
+
+
+def with_session(func):
+    """Decorator for methods that require the session in the context """
```
Comment (Vegter) [inline src/gobuploadservice/storage/storage_handler.py]: Docstrings missing

File: src/gobuploadservice/storage/storage_handler.py [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+        table.create(self.engine, checkfirst=True)
+
+    @property
+    def DbEvent(self):
+        return self.base.classes.event
+
+    @property
+    def DbEntity(self):
+        return getattr(self.base.classes, self.metadata.entity)
+
```
Comment (Vegter) [inline src/gobuploadservice/storage/storage_handler.py]: db or storage? I would recommend consistent naming.
If the storage handler encapsulates the storage (being a db, file, json, collections of csv's) the name db is a little confusing. Actually, the data can be stored wherever what structure. Following this reasoning, why not give the storage handler simply properties event and entity?

File: src/gobuploadservice/storage/storage_handler.py [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+            def __exit__(ctx, type, value, traceback):
+                self.session.commit()
+                self.session.close()
+                self.session = None
+
+        return session_context()
+
+    @with_session
+    def get_current_ids(self):
+        return self.session.query(self.DbEntity._source_id).filter_by(_source=self.metadata.source,
```
Comment (Vegter) [inline src/gobuploadservice/storage/storage_handler.py]: docstrings?

File: src/gobuploadservice/storage/storage_handler.py [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+                                                                      _date_deleted=None).all()
+
+    @with_session
+    def get_entity_or_none(self, entity_id, with_deleted=False):
+        entity_query = self.session.query(self.DbEntity).filter_by(_source=self.metadata.source, _source_id=entity_id)
```
Comment (Vegter) [inline src/gobuploadservice/storage/storage_handler.py]: docstrings

File: src/gobuploadservice/storage/storage_handler.py [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+        if not with_deleted:
+            entity_query = entity_query.filter_by(_date_deleted=None)
+
+        return entity_query.one_or_none()
+
+    @with_session
+    def add_event_to_db(self, event):
+        entity = build_db_event(self.DbEvent, event, self.metadata)
```
Comment (Vegter) [inline src/gobuploadservice/storage/storage_handler.py]: docstrings, add_event

File: src/gobuploadservice/storage/storage_handler.py [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+        self.session.add(entity)
+
+    def get_entity_for_update(self, entity_id, source_id, gob_event):
+        entity = self.get_entity_or_none(source_id, with_deleted=True)
```
Comment (Vegter) [inline src/gobuploadservice/storage/storage_handler.py]: docstrings

File: src/gobuploadservice/update.py [hunk 13]
```diff
@@ -28,53 +16,30 @@ def full_update(msg):
...
-                                                   _source_id=source_id).one_or_none()
-
-        # Make sure an entity is available when the event requires an addition.
-        if gob_event.is_add_new:
-            if entity is None:
-                # create a new Entity
-                entity = DbEntity(_source_id=source_id, _source=metadata.source)
-                setattr(entity, metadata.id_column, entity_id)
-                session.add(entity)
+    db = GOBStorageHandler(metadata)
```
Comment (Vegter) [inline src/gobuploadservice/update.py]: storage = GOB...

Comment (dolhuis) [reply src/gobuploadservice/update.py]: Done

File: src/gobuploadservice/compare.py (hunk 1 above)
Comment (dolhuis) [reply src/gobuploadservice/compare.py]: Done

File: src/gobuploadservice/storage/storage_handler.py (hunk 2 above)
Comment (dolhuis) [reply src/gobuploadservice/storage/storage_handler.py]: Done

File: src/Dockerfile (hunk 3 above)
Comment (dolhuis) [reply src/Dockerfile]: Removed it, it was not necessary

File: src/gobuploadservice/compare.py (hunk 5 above)
Comment (dolhuis) [reply src/gobuploadservice/compare.py]: Done

File: src/gobuploadservice/storage/storage_handler.py (hunk 6 above)
Comment (dolhuis) [reply src/gobuploadservice/storage/storage_handler.py]: Done

File: src/gobuploadservice/storage/storage_handler.py (hunk 9 above)
Comment (dolhuis) [reply src/gobuploadservice/storage/storage_handler.py]: Done

File: src/gobuploadservice/storage/storage_handler.py (hunk 10 above)
Comment (dolhuis) [reply src/gobuploadservice/storage/storage_handler.py]: Done

File: src/gobuploadservice/storage/storage_handler.py (hunk 11 above)
Comment (dolhuis) [reply src/gobuploadservice/storage/storage_handler.py]: Done

File: src/gobuploadservice/storage/storage_handler.py (hunk 12 above)
Comment (dolhuis) [reply src/gobuploadservice/storage/storage_handler.py]: Done

File: src/gobuploadservice/storage/storage_handler.py (hunk 7 above)
Comment (dolhuis) [reply src/gobuploadservice/storage/storage_handler.py]: Added more documentation

**Pull request merged**